x-pack/plugins/security_solution/server/endpoint/services/lookup_agent.ts

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SearchResponse } from 'elasticsearch';
+import { HostMetadata } from '../../../common/endpoint/types';
+import { SecuritySolutionRequestHandlerContext } from '../../types';
+import { getESQueryHostMetadataByIDs } from '../routes/metadata/query_builders';
+import { EndpointAppContext } from '../types';
+
+export async function GetAgentIDsForEndpoints(
+  endpointIDs: string[],
+  requestHandlerContext: SecuritySolutionRequestHandlerContext,
+  endpointAppContext: EndpointAppContext
+): Promise<string[]> {
+  const queryStrategy = await endpointAppContext.service
+    ?.getMetadataService()
+    ?.queryStrategy(requestHandlerContext.core.savedObjects.client);
+
+  const query = getESQueryHostMetadataByIDs(endpointIDs, queryStrategy!);
+  const esClient = requestHandlerContext.core.elasticsearch.client.asCurrentUser;
+  // @ts-expect-error
+  const { body }: { body: SearchResponse<HostMetadata> } = await esClient.search(query);
```

Comment:
Which cain of ts error do you have here?

Reply:
two here that I need to fix. One on `body`:
```
Type 'ApiResponse<SearchResponse<ImmutableObject<{ '@timestamp': number; event: { created: number; kind: string; id: string; category: string[]; type: string[]; module: string; action: string; dataset: string; }; elastic: { agent: { id: string; }; }; Endpoint: { ...; }; agent: { ...; }; host: Host; data_stream: DataStream...' is not assignable to type '{ body: SearchResponse<ImmutableObject<{ '@timestamp': number; event: { created: number; kind: string; id: string; category: string[]; type: string[]; module: string; action: string; dataset: string; }; elastic: { agent: { id: string; }; }; Endpoint: { ...; }; agent: { ...; }; host: Host; data_stream: DataStream; }>...'.
  The types of 'body._shards.skipped' are incompatible between these types.
    Type 'number | undefined' is not assignable to type 'number'.
      Type 'undefined' is not assignable to type 'number'.
```
`body._shards.skipped` is from an ES search response, I don't know why it's complaining about that


and another on `query`:
```
Argument of type '{ body: { query: { bool: { filter: { bool: { should: ({ terms: { 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }; } | { terms: { 'HostDetails.agent.id': string[]; 'agent.id'?: undefined; }; })[]; }; }[]; }; }; sort: ({ ...; } | { ...; })[]; }; index: string; }' is not assignable to parameter of type 'SearchRequest'.
  The types of 'body.query.bool.filter' are incompatible between these types.
    Type '{ bool: { should: ({ terms: { 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }; } | { terms: { 'HostDetails.agent.id': string[]; 'agent.id'?: undefined; }; })[]; }; }[]' is not assignable to type 'QueryContainer | QueryContainer[] | undefined'.
      Type '{ bool: { should: ({ terms: { 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }; } | { terms: { 'HostDetails.agent.id': string[]; 'agent.id'?: undefined; }; })[]; }; }[]' is not assignable to type 'QueryContainer[]'.
        Type '{ bool: { should: ({ terms: { 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }; } | { terms: { 'HostDetails.agent.id': string[]; 'agent.id'?: undefined; }; })[]; }; }' is not assignable to type 'QueryContainer'.
          The types of 'bool.should' are incompatible between these types.
            Type '({ terms: { 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }; } | { terms: { 'HostDetails.agent.id': string[]; 'agent.id'?: undefined; }; })[]' is not assignable to type 'QueryContainer | QueryContainer[] | undefined'.
              Type '({ terms: { 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }; } | { terms: { 'HostDetails.agent.id': string[]; 'agent.id'?: undefined; }; })[]' is not assignable to type 'QueryContainer[]'.
                Type '{ terms: { 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }; } | { terms: { 'HostDetails.agent.id': string[]; 'agent.id'?: undefined; }; }' is not assignable to type 'QueryContainer'.
                  Type '{ terms: { 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }; }' is not assignable to type 'QueryContainer'.
                    Types of property 'terms' are incompatible.
                      Type '{ 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }' is not assignable to type 'NamedQueryKeys<string[] | TermsQuery | number[]> | { [property: string]: string[] | TermsQuery | number[]; } | undefined'.
                        Type '{ 'agent.id': string[]; 'HostDetails.agent.id'?: undefined; }' is not assignable to type '{ [property: string]: string[] | TermsQuery | number[]; }'.
                          Property ''HostDetails.agent.id'' is incompatible with index signature.
                            Type 'undefined' is not assignable to type 'string[] | TermsQuery | number[]'.ts(2345)
```

this is also a fun one. It's trying to validate the query being executed, and it doesn't like that we have a `should` testing two different fields. I don't actually know why it's calling it `undefined`. The query itself is here:

```ts
export function getESQueryHostMetadataByIDs(
  agentIDs: string[],
  metadataQueryStrategy: MetadataQueryStrategy
) {
  return {
    body: {
      query: {
        bool: {
          filter: [
            {
              bool: {
                should: [
                  { terms: { 'agent.id': agentIDs } },
                  { terms: { 'HostDetails.agent.id': agentIDs } },
                ],
              },
            },
          ],
        },
      },
      sort: MetadataSortMethod,
    },
    index: metadataQueryStrategy.index,
  };
}
```

as you can see, `agentIDs` is not nullable, and passed to both. But the implicit return type of that big structure throws in the `undefined`

Reply:
🤯 
It still happen if you do a type casting?

Reply:
The `body` one seems very silly. Because it seems to be erroring on only pure ES types:

`esClient.search` returns  `TransportRequestPromise<ApiResponse<T.SearchResponse<TDocument>, TContext>>`

we are awaiting, so lets unwrap to `ApiResponse<T.SearchResponse<TDocument>, TContext>` and `body` is the destructured property to `SearchResponse<TDocument>` where `TDocument` is our `HostMetadata`.

Ok.. so the type should be `SearchResponse<HostMetadata>` eh?

And `queryResponseToHostListResult` accepts a type `SearchResponse<HostMetadata>` as the input param.

So I'm really just trying to pass it over.

But it's complaining about some internal field. Oye.

As for the search query, I know that query is valid, it works. I don't need it type checked.


Open to what the solution here should be, considering they both seem to be silly situations I don't really care for type checking here anyway